### PR TITLE
Porous Flow initial 

### PR DESCRIPTION
**Pull request by TaraLaForce**
Refs #6740

### CODE REVIEW

File: modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+
```
Comment (dschwen) [inline modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h]: empty lines

File: modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ */
+class PorFlowComponentMassTimeDerivative : public TimeKernel
+{
+public:
+
+  PorFlowComponentMassTimeDerivative(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual();
+
```
Comment (dschwen) [inline modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h]: Using empty lines in this manner does not seem particularly useful to me. IMO they should be used sparingly to visually separate blocks that go together. In this particular file I'd suggest adding `///` comments to document the members as well.

File: modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  const MaterialProperty<std::vector<std::vector<std::vector<Real> > > > & _dmass_frac_dvar;
+
+  /**
+   * Derivative of residual with respect to PorFlow variable number pvar
+   * This is used by both computeQpJacobian and computeQpOffDiagJacobian
+   * @param wrt_num take the derivative of the residual wrt this PorFlow variable
+   */
+  Real computeQpJac(unsigned int pvar);
+
```
Comment (dschwen) [inline modules/porous_flow/include/kernels/PorFlowComponentMassTimeDerivative.h]: empty line

File: modules/porous_flow/include/materials/PorFlowMaterial2PhasePS.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+
```
Comment (dschwen) [inline modules/porous_flow/include/materials/PorFlowMaterial2PhasePS.h]: empty lines

File: modules/porous_flow/include/materials/PorFlowMaterial2PhasePS.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  MaterialProperty<std::vector<Real> > & _saturation;
+
+  /// old value of saturation of the phases
+  MaterialProperty<std::vector<Real> > & _saturation_old;
+
+  /// d(porepressure)/d(porflow variable)
+  MaterialProperty<std::vector<std::vector<Real> > > & _dsaturation_dvar;
+
+  virtual void initQpStatefulProperties();
+  virtual void computeQpProperties();
```
Comment (dschwen) [inline modules/porous_flow/include/materials/PorFlowMaterial2PhasePS.h]: The de facto standard seems to be to have member methods first and then member variables,

File: modules/porous_flow/src/materials/PorFlowMaterialDensityBuilder.C [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    _num_phases(getParam<unsigned int>("num_phases")),
+    _porflow_name_UO(getUserObject<PorFlowVarNames>("PorFlowVarNames_UO")),
+
+    _density(declareProperty<std::vector<Real> >("PorFlow_fluid_phase_density")),
+    _density_old(declarePropertyOld<std::vector<Real> >("PorFlow_fluid_phase_density")),
+    _ddensity_dvar(declareProperty<std::vector<std::vector<Real> > >("dPorFlow_fluid_phase_density_dvar"))
+{
+  _phase_density.resize(_num_phases);
+  _dphase_density_dvar.resize(_num_phases);
```
Comment (dschwen) [inline modules/porous_flow/src/materials/PorFlowMaterialDensityBuilder.C]: these should be moved into the initializer lists

File: modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ * of mass fractions from the individual mass fraction variables
+ */
+class PorFlowMaterialMassFractionBuilder : public DerivativeMaterialInterface<Material>
+{
+public:
+  PorFlowMaterialMassFractionBuilder(const InputParameters & parameters);
+
+protected:
+
+  unsigned int _num_phases;
```
Comment (dschwen) [inline modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h]: Should be `const` (no empty line after the access qualifiers)

File: modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+  unsigned int _num_components;
+
+  /// The variable names UserObject for the Porous-Flow variables
+  const PorFlowVarNames & _porflow_name_UO;
```
Comment (dschwen) [inline modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h]: Sooo.... there was that discussion about the `PorFlow` prefix... I know that it is quite bikeshed, but what are you going to go with now?

File: modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  unsigned int _num_passed_mf_vars;
+
+  std::vector<unsigned int> _mf_vars_num;
+  std::vector<const VariableValue *> _mf_vars;
+
+  virtual void initQpStatefulProperties();
+  virtual void computeQpProperties();
+
+ private:
+  void build_mass_frac(unsigned int qp);
```
Comment (dschwen) [inline modules/porous_flow/include/materials/PorFlowMaterialMassFractionBuilder.h]: Should be named `buildMassFraction()`

File: modules/porous_flow/include/userobjects/PorFlowVarNames.h [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  unsigned int num_v() const;
+
+  /**
+   * the PorFlow variable number
+   * @param moose_var_num the MOOSE variable number
+   * eg if porflow_vars = 'pwater pgas', and the variables in
+   * the simulation are 'energy pwater pgas shape'
+   * then porflow_var_num(2) = 1
+   */
+  unsigned int porflow_var_num(unsigned int moose_var_num) const;
```
Comment (dschwen) [inline modules/porous_flow/include/userobjects/PorFlowVarNames.h]: all member function names should be camelCased

File: modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+{
+  InputParameters params = validParams<TimeKernel>();
+  params.addParam<unsigned int>("component_index", 0, "The index corresponding to the component for this kernel");
+  params.addRequiredParam<UserObjectName>("PorFlowVarNames_UO", "The UserObject that holds the list of Porous-Flow variable names.");
+  params.addClassDescription("Component mass derivative wrt time for component given by component_index");
+  return params;
+}
+
+PorFlowComponentMassTimeDerivative::PorFlowComponentMassTimeDerivative(const InputParameters & parameters) :
+  TimeKernel(parameters),
```
Comment (dschwen) [inline modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C]: initializer indent is 4 spaces

File: modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+/// Note that this kernel lumps the mass terms to the nodes, so that there is no mass at the qp's.
+Real
+PorFlowComponentMassTimeDerivative::computeQpResidual()
+{
+  mooseAssert(_component_index < _mass_frac[_i][0].size(), "PorFlowComponentMassTimeDerivative: component_index is given as " << _component_index << " which must be less than the number of fluid components described by the mass-fraction matrix, which is " << _mass_frac[_i][0].size());
+  unsigned int num_phases = _fluid_density[_i].size();
+  mooseAssert(num_phases == _fluid_saturation[_i].size(), "PorFlowComponentMassTimeDerivative: Size of fluid density = " << num_phases << " size of fluid saturation = " << _fluid_saturation[_i].size() << " but both these must be equal to the number of phases in the system");
+
+  Real mass = 0.;
```
Comment (dschwen) [inline modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C]: `0.0`

File: modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+   }
+
+  return _test[_i][_qp] * (_porosity[_i]*mass - _porosity_old[_i]*mass_old)/_dt;
+}
+
+Real
+PorFlowComponentMassTimeDerivative::computeQpJacobian()
+{
+  /// If the variable is not a PorFlow variable (very unusual), the diag Jacobian terms are 0
+  if (!(_var_is_porflow_var))
```
Comment (dschwen) [inline modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C]: inner set of parens is unnecessary

File: modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+  /// As mass is lumped to the nodes, only non-zero terms are for _i==_j
+  if (_i != _j)
+    return 0.0;
+
+  unsigned int num_phases = _fluid_density[_i].size();
+
+  Real dmass = 0.0;
+  for (unsigned ph = 0; ph < num_phases; ++ph)
+  {
+    dmass += _dfluid_density_dvar[_i][ph][pvar]*_fluid_saturation[_i][ph]*_mass_frac[_i][ph][_component_index]*_porosity[_i];
```
Comment (dschwen) [inline modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C]: operator spacing (also the repeated use of `+=` should probably be replaced by just letting the term flow over multiple lines and `+` operators)

File: modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    dmass += _fluid_density[_i][ph]*_dfluid_saturation_dvar[_i][ph][pvar]*_mass_frac[_i][ph][_component_index]*_porosity[_i];
+    dmass += _fluid_density[_i][ph]*_fluid_saturation[_i][ph]*_dmass_frac_dvar[_i][ph][_component_index][pvar]*_porosity[_i];
+    dmass += _fluid_density[_i][ph]*_fluid_saturation[_i][ph]*_mass_frac[_i][ph][_component_index]*_dporosity_dvar[_i][pvar];
+  }
+  return _test[_i][_qp]*dmass/_dt;
```
Comment (dschwen) [inline modules/porous_flow/src/kernels/PorFlowComponentMassTimeDerivative.C]: more operator spacing

File: modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+
+
+#include "PorFlowMaterial2PhasePS.h"
+
+
+
```
Comment (dschwen) [inline modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C]: rm 4 of these empty lines

File: modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+  _porepressure[_qp].resize(2);
+  _dporepressure_dvar[_qp].resize(2);
+  _porepressure_old[_qp].resize(2);
+
+  _saturation[_qp].resize(2);
+  _dsaturation_dvar[_qp].resize(2);
+  _saturation_old[_qp].resize(2);
+
+  /*
+   *  YAQI HACK !!
```
Comment (dschwen) [inline modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C]: LOL

File: modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+   *       but only if i keep pc=constant, which is probably unphysical, so i'll
+   *       keep the following computations here for modification with
+   *       a physical pc later.
+   */
+  // prepare the derivative matrix with zeroes
+  for (unsigned phase = 0; phase < 2; ++phase)
+    _dporepressure_dvar[_qp][phase].assign(_porflow_name_UO.num_v(), 0.0);
+
+  // _porepressure is only dependent on _phase0_porepressure, and its derivative is 1
+  if (!(_porflow_name_UO.not_porflow_var(_phase0_porepressure_varnum)))
```
Comment (dschwen) [inline modules/porous_flow/src/materials/PorFlowMaterial2PhasePS.C]: remove extra parens

File: modules/porous_flow/src/materials/PorFlowMaterialMassFractionBuilder.C [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+    _mass_frac(declareProperty<std::vector<std::vector<Real> > >("PorFlow_mass_frac")),
+    _mass_frac_old(declarePropertyOld<std::vector<std::vector<Real> > >("PorFlow_mass_frac")),
+    _dmass_frac_dvar(declareProperty<std::vector<std::vector<std::vector<Real> > > >("dPorFlow_mass_frac_dvar")),
+
+    _num_passed_mf_vars(coupledComponents("mass_fraction_vars"))
+{
+  if (_num_passed_mf_vars != _num_phases*(_num_components - 1))
+    mooseError("PorFlowMaterialMassFractionBuilder: The number of mass_fraction_vars is " << _num_passed_mf_vars << " which must be equal to num_phases (" << _num_phases << ") multiplied by num_components-1 (" << _num_components - 1 << ")");
+
+  _mf_vars.resize(_num_passed_mf_vars);
```
Comment (dschwen) [inline modules/porous_flow/src/materials/PorFlowMaterialMassFractionBuilder.C]: move all these resizes into the initializer list
`    _mf_vars(_num_passed_mf_vars),`

File: modules/porous_flow/src/userobjects/PorFlowVarNames.C [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+  _pf_var_num.resize(max_moose_var_num_seen + 1);
+  for (unsigned int i = 0; i < max_moose_var_num_seen + 1; ++i)
+    _pf_var_num[i] = _num_v; // NOTE: indicates that i is not a richards variable
+  for (unsigned int i = 0; i < _num_v; ++i)
+    _pf_var_num[_moose_var_num[i]] = i;
+}
+
+void
+PorFlowVarNames::initialize()
+{}
```
Comment (dschwen) [inline modules/porous_flow/src/userobjects/PorFlowVarNames.C]:
```
{
}
```

**Comment by WilkAndy** [comment]
Please close this PR.  It is outdated and i'll submit a new one shortly.

**Comment by permcody** [comment]
Closing in favor of #6850

**Pull request closed without merging**